CONTRIBUTING.rst

```diff
@@ -380,6 +380,56 @@ pull requests from other contributors yourself after having reviewed
 them.
 
 
+Merge/squash guidelines
+-----------------------
+
+When a PR is approved and ready to be integrated to the `main` branch, one has the option to *merge* the commits unchanged, or *squash* all the commits into a single commit.
+
+Here is some guidelines on how to proceeed, based on the state of the PR commit history:
+
+1. Miscellaneous commits:
+
+   * `Implement X`
+   * `Fix test_a`
+   * `Add myself to AUTHORS`
+   * `!fixup Fix test_a`
+   * `Update tests/test_integration.py`
+   * `Update tests/test_integration.py`
+
+   In this case, prefer to use the **Squash** merge strategy: the commit history is a bit messy (not in a derrogatory way, often one just commits changes because they know the changes will eventually be squashed together), so squashing everything into a single commit is best. Prefer to also edit the default GitHub message to add more details and improve on it.
+
+2. Separate commits related to the same topic:
+
+   * `Implement X`
+   * `Add myself to AUTHORS`
+   * `Update CHANGELOG for X`
+
+   In this case, prefer to use the **Squash** merge strategy: while the commit history is not "messy" as in the example above, the individual commits do not bring much value overall, specially when looking at the changes a few months/years down the line.
+
+3. Separate commits, each with their own topic (refactorings, renames, etc), but still have a larger topic/purpose.
+
+   * `Refactor class X in preparation for feature Y`
+   * `Remove unused method`
+   * `Implement feature Y`
+
+   In this case, prefer to use the **Merge** strategy: each commit is valuable on its own, even if they serve a common topic overall. Looking at the history later, it is useful to have the removal of the unused method separately on its own commit, along with more information (such as how it became unused in the first place).
+
+4. Separate commits, each with their own topic, but without a larger topic/purpose other than improve the code base (using more modern techniques, improve typing, removing clutter, etc).
```

Comment:
Should we note somewhere that such PRs might not be backportable? (since those are squashed)

Reply:
Not sure what you mean, can you elaborate? The example 4 above the PR is merged, so individual commits might still be backported.

Reply:
@nicoddemus there is no mechanism for automatically backporting individual commits. They can only be backported together as a whole this way.

Reply:
I'm aware, not sure what you meant by your first comment? That example 4 above does not even mention backports...

Reply:
It doesn't mention backports, but it says that unrelated commits can be in one PR. And applying backport labels to such PRs will inevitably offer backports where those are squashed.

Reply:
I see... want to make a suggestion then?

Reply:
I'd normally suggest not making such PRs. Asking people to *remember* to pick out individual commits when backporting is an extra maintenance burden. I don't see how that could work realistically.

Reply:
OTOH, maybe this could be a FR for Patchback — have a label for communicating such an intent.

Reply:
Sorry I meant a direct suggestion to the text (as in "```suggestion"). 😁

Reply:
I don't have a nice phrasing in my mind for this right now. Also, I cannot imagine such a PR existing with a good description. Still, if it does have a good description, maybe it's not that bad if a bot would smash everything together in stable branches.